lib/engine-universal/src/artifact.rs

```diff
@@ -1,402 +1,201 @@
-                })
-            }
-            None => None,
-        };
-
-        // Make all code compiled thus far executable.
-        inner_engine.publish_compiled_code();
-
-        inner_engine.publish_eh_frame(eh_frame)?;
-
-        let finished_function_lengths = finished_functions
-            .values()
-            .map(|extent| extent.length)
-            .collect::<PrimaryMap<LocalFunctionIndex, usize>>()
-            .into_boxed_slice();
-        let finished_functions = finished_functions
-            .values()
-            .map(|extent| extent.ptr)
-            .collect::<PrimaryMap<LocalFunctionIndex, FunctionBodyPtr>>()
-            .into_boxed_slice();
-        let finished_function_call_trampolines =
-            finished_function_call_trampolines.into_boxed_slice();
-        let finished_dynamic_function_trampolines =
-            finished_dynamic_function_trampolines.into_boxed_slice();
-        let signatures = signatures.into_boxed_slice();
-        let func_data_registry = inner_engine.func_data().clone();
-
-        Ok(Self {
-            serializable,
-            finished_functions,
-            finished_function_call_trampolines,
-            finished_dynamic_function_trampolines,
-            signatures,
-            frame_info_registration: Mutex::new(None),
-            finished_function_lengths,
-            func_data_registry,
+    /// Return the extents of the specified local function.
+    pub fn function_extent(&self, index: LocalFunctionIndex) -> Option<FunctionExtent> {
+        let func = self.functions.get(index)?;
+        Some(FunctionExtent {
+            address: func.body,
+            length: usize::try_from(func.length).unwrap(),
         })
     }
 
-    /// Get the default extension when serializing this artifact
-    pub fn get_default_extension(_triple: &Triple) -> &'static str {
-        // `.wasmu` is the default extension for all the triples. It
-        // stands for “Wasm Universal”.
-        "wasmu"
+    /// Return the engine instance this artifact is loaded into.
+    pub fn engine(&self) -> &crate::UniversalEngine {
+        &self.engine
     }
 }
 
 impl Artifact for UniversalArtifact {
-    fn module(&self) -> Arc<ModuleInfo> {
-        self.serializable.compile_info.module.clone()
-    }
-
-    fn module_ref(&self) -> &ModuleInfo {
-        &self.serializable.compile_info.module
-    }
-
-    fn module_mut(&mut self) -> Option<&mut ModuleInfo> {
-        Arc::get_mut(&mut self.serializable.compile_info.module)
-    }
-
-    fn register_frame_info(&self) {
-        let mut info = self.frame_info_registration.lock().unwrap();
+    unsafe fn instantiate(
+        self: Arc<Self>,
+        tunables: &dyn Tunables,
+        resolver: &dyn Resolver,
+        host_state: Box<dyn std::any::Any>,
+        config: wasmer_types::InstanceConfig,
+    ) -> Result<InstanceHandle, Box<dyn std::error::Error + Send + Sync>> {
```

Comment:
Hm, the error handling here looks suspicious. I see this has at least two error paths:

* unresolved import
* failure to allocate memory

At the nearcore side, we for sure want to handle these two erros differently. Unrosolved import is a normal outcome of a contract execution, while failure to allocate memory is a non-deterministic error, which should crash the node.